File: db4on/core/src/core/com/db4o/EventDispatcher.cs
namespace com.db4o
{
	/// <exclude></exclude>
	public sealed class EventDispatcher
	{
		private static readonly string[] events = { "objectCanDelete", "objectOnDelete", 
			"objectOnActivate", "objectOnDeactivate", "objectOnNew", "objectOnUpdate", "objectCanActivate"
			, "objectCanDeactivate", "objectCanNew", "objectCanUpdate" };

		internal const int CAN_DELETE = 0;

		internal const int DELETE = 1;

		internal const int SERVER_COUNT = 2;

		internal const int ACTIVATE = 2;

		internal const int DEACTIVATE = 3;

		internal const int NEW = 4;

		public const int UPDATE = 5;

		internal const int CAN_ACTIVATE = 6;

		internal const int CAN_DEACTIVATE = 7;

		internal const int CAN_NEW = 8;

		internal const int CAN_UPDATE = 9;

		internal const int COUNT = 10;

		private readonly com.db4o.reflect.ReflectMethod[] methods;

		private EventDispatcher(com.db4o.reflect.ReflectMethod[] methods_)
		{
			methods = methods_;
		}

		internal bool Dispatch(com.db4o.YapStream stream, object obj, int eventID)
		{
			if (methods[eventID] != null)
			{
				object[] parameters = new object[] { stream };
				try
				{
					object res = methods[eventID].Invoke(obj, parameters);
					if (res != null && res is bool)
					{
						return ((bool)res);
					}
				}
				catch
				{
				}
			}
			return true;
		}

		internal static com.db4o.EventDispatcher ForClass(com.db4o.YapStream a_stream, com.db4o.reflect.ReflectClass
			 classReflector)
		{
			if (a_stream == null || classReflector == null)
			{
				return null;
			}
			com.db4o.EventDispatcher dispatcher = null;
			int count = 0;
			if (a_stream.ConfigImpl().Callbacks())
			{
				count = COUNT;
			}
			else
			{
				if (a_stream.ConfigImpl().IsServer())
				{
					count = SERVER_COUNT;
				}
			}
			if (count > 0)
			{
				com.db4o.reflect.ReflectClass[] parameterClasses = { a_stream.i_handlers.ICLASS_OBJECTCONTAINER
					 };
				com.db4o.reflect.ReflectMethod[] methods = new com.db4o.reflect.ReflectMethod[COUNT
					];
				for (int i = COUNT - 1; i >= 0; i--)
				{
					try
					{
						com.db4o.reflect.ReflectMethod method = classReflector.GetMethod(events[i], parameterClasses
							);
						if (null == method)
						{
							method = classReflector.GetMethod(ToPascalCase(events[i]), parameterClasses);
						}
						methods[i] = method;
						if (dispatcher == null)
						{
							dispatcher = new com.db4o.EventDispatcher(methods);
						}
					}
					catch
					{
					}
				}
			}
			return dispatcher;
		}

		private static string ToPascalCase(string name)
		{
			return j4o.lang.JavaSystem.Substring(name, 0, 1).ToUpper() + j4o.lang.JavaSystem.Substring
				(name, 1);
		}
	}
}

File: db4oj/core/src/com/db4o/internal/EventDispatcher.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.internal;

import com.db4o.foundation.*;
import com.db4o.reflect.*;


/** @exclude */
public final class EventDispatcher {
	private static final String[] events = {
		"objectCanDelete",
		"objectOnDelete", 
		"objectOnActivate", 
		"objectOnDeactivate",
		"objectOnNew",
		"objectOnUpdate",
		"objectCanActivate",
		"objectCanDeactivate",
		"objectCanNew",
		"objectCanUpdate"
	};
	
	static final int CAN_DELETE = 0;
	static final int DELETE = 1;
	static final int SERVER_COUNT = 2;
	static final int ACTIVATE = 2;
	static final int DEACTIVATE = 3;
	static final int NEW = 4;
	public static final int UPDATE = 5;
	static final int CAN_ACTIVATE = 6;
	static final int CAN_DEACTIVATE = 7;
	static final int CAN_NEW = 8;
	static final int CAN_UPDATE = 9;
	static final int COUNT = 10;
	
	private final ReflectMethod[] methods;
	
	private EventDispatcher(ReflectMethod[] methods_){
		methods = methods_;
	}
	
	boolean dispatch(Transaction trans, Object obj, int eventID) {
		if (methods[eventID] == null) {
			return true;
		}
		Object[] parameters = new Object[] { trans.objectContainer()};
		ObjectContainerBase container = trans.container();
		int stackDepth = container.stackDepth();
		int topLevelCallId = container.topLevelCallId();
		container.stackDepth(0);
		try {
			Object res = methods[eventID].invoke(obj, parameters);
			if (res instanceof Boolean) {
				return ((Boolean) res).booleanValue();
			}
		}  finally {
			container.stackDepth(stackDepth);
			container.topLevelCallId(topLevelCallId);
		}
		return true;
	}
	
	static EventDispatcher forClass(ObjectContainerBase container, ReflectClass classReflector){
        
        if(container == null || classReflector == null){
           throw new ArgumentNullException();
        }
        
		EventDispatcher dispatcher = null;
	    int count = 0;
	    if(container.configImpl().callbacks()){
	        count = COUNT;
	    }else if(container.configImpl().isServer()){
	        count = SERVER_COUNT;
	    }
	    if(count > 0){
			ReflectClass[] parameterClasses = {container._handlers.ICLASS_OBJECTCONTAINER};
			ReflectMethod[] methods = new ReflectMethod[COUNT];
			for (int i = COUNT - 1; i >= 0; i--) {
				ReflectMethod method = classReflector.getMethod(events[i],
						parameterClasses);
				if (null == method) {
					method = classReflector.getMethod(toPascalCase(events[i]),
							parameterClasses);
				}
				if (method != null) {
					methods[i] = method;
					if (dispatcher == null) {
						dispatcher = new EventDispatcher(methods);
					}
				}
			}
	    }
        
		return dispatcher;
	}

	private static String toPascalCase(String name) {
		return name.substring(0, 1).toUpperCase() + name.substring(1);
	}

    public boolean hasEventRegistered(int eventID){
        return methods[eventID] != null;
    }
}
